README.md
# timestampmath

Workbench module that compares and adjusts timestamps

# Development

Do the Development Dance:

1. Run `tox` and make sure it works
2. Write a failing test to `tests/`
3. Run `tox` to test it fails
4. Fix the code until `tox` passes
5. Submit a pull request

# Deployment

1. Write an entry to `CHANGELOG.md`
2. Run `tox` one last time
3. `git push`


CHANGELOG.md
2021-05-05.01
-------------

* Add "startof" for rounding hour/minute/second/millisecond/microsecond.

2020-10-02.01
-------------

* (internal) rename "datetime" to "timestamp"

2020-09-30.01
-------------

* Fix error calculating difference in zero-row tables.

2020-09-29.01
-------------

* Initial release: `difference`, `minimum` and `maximum`.

`difference` uses floating-point math, not integer math. In FP mode, it
probably shouldn't support month/year differences: that would be non-standard
math. But someday, we may consider an integer mode to allow Gregorian-calendar
math in the user's timezone.


tests/test_render.py
from datetime import datetime as dt
from pathlib import Path

import pyarrow as pa
from cjwmodule.arrow.testing import assert_result_equals, make_column, make_table
from cjwmodule.arrow.types import ArrowRenderResult
from cjwmodule.spec.testing import param_factory
from cjwmodule.testing.i18n import i18n_message
from cjwmodule.types import RenderError

from timestampmath import render_arrow_v1 as render

P = param_factory(Path(__file__).parent.parent / "timestampmath.yaml")


def test_maximum_no_colnames():
    assert_result_equals(
        render(
            make_table(make_column("A", [1, 2, 3], pa.timestamp(unit="ns"))),
            P(operation="maximum", colnames=[], outcolname="B"),
        ),
        ArrowRenderResult(
            make_table(make_column("A", [1, 2, 3], pa.timestamp(unit="ns"))),
        ),
    )


def test_maximum_no_outcolname():
    assert_result_equals(
        render(
            pa.table({"A": pa.array([1, 2, 3], pa.timestamp(unit="ns"))}),
            P(operation="maximum", colnames=["A"], outcolname=""),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, 2, 3], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_maximum_one_colname():
    assert_result_equals(
        render(
            pa.table({"A": pa.array([1, 2, 3], pa.timestamp(unit="ns"))}),
            P(operation="maximum", colnames=["A"], outcolname="B"),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, 2, 3], pa.timestamp(unit="ns")),
                    "B": pa.array([1, 2, 3], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_maximum_multiple_colnames():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, 2, 3], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3, 1], pa.timestamp(unit="ns")),
                    "C": pa.array([3, 1, 2], pa.timestamp(unit="ns")),
                }
            ),
            P(operation="maximum", colnames=["A", "B", "C"], outcolname="D"),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, 2, 3], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3, 1], pa.timestamp(unit="ns")),
                    "C": pa.array([3, 1, 2], pa.timestamp(unit="ns")),
                    "D": pa.array([3, 3, 3], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_maximum_reuse_outcolname():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, 2, 3], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3, 1], pa.timestamp(unit="ns")),
                }
            ),
            P(operation="maximum", colnames=["A", "B"], outcolname="A"),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "B": pa.array([2, 3, 1], pa.timestamp(unit="ns")),
                    "A": pa.array([2, 3, 3], pa.timestamp(unit="ns")),  # added to end
                }
            ),
        ),
    )


def test_maximum_zero_chunks():
    assert_result_equals(
        render(
            pa.table({"A": pa.array([], pa.timestamp(unit="ns"))}),
            P(operation="maximum", colnames=["A"], outcolname="B"),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([], pa.timestamp(unit="ns")),
                    "B": pa.array([], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_maximum_nulls():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, None, 3, None], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3, None, None], pa.timestamp(unit="ns")),
                }
            ),
            P(operation="maximum", colnames=["A", "B"], outcolname="C"),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, None, 3, None], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3, None, None], pa.timestamp(unit="ns")),
                    "C": pa.array([2, 3, 3, None], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_minimum():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, None, 3, None], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3, None, None], pa.timestamp(unit="ns")),
                }
            ),
            P(operation="minimum", colnames=["A", "B"], outcolname="C"),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, None, 3, None], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3, None, None], pa.timestamp(unit="ns")),
                    "C": pa.array([1, 3, 3, None], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_difference_no_colnames():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
            P(
                operation="difference",
                colname1="",
                colname2="",
                unit="day",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_difference_no_colname1():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
            P(
                operation="difference",
                colname1="",
                colname2="B",
                unit="day",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_difference_no_colname2():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="",
                unit="day",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_difference_no_outcolname():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="day",
                outcolname="",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([1, 2], pa.timestamp(unit="ns")),
                    "B": pa.array([2, 3], pa.timestamp(unit="ns")),
                }
            ),
        ),
    )


def test_difference_no_chunks():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.chunked_array([], pa.timestamp(unit="ns")),
                    "B": pa.chunked_array([], pa.timestamp(unit="ns")),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="day",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.chunked_array([], pa.timestamp(unit="ns")),
                    "B": pa.chunked_array([], pa.timestamp(unit="ns")),
                    "C": pa.chunked_array([], pa.float64()),
                }
            ),
        ),
    )


def test_difference_no_chunks_nanoseconds():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.chunked_array([], pa.timestamp(unit="ns")),
                    "B": pa.chunked_array([], pa.timestamp(unit="ns")),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="nanosecond",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.chunked_array([], pa.timestamp(unit="ns")),
                    "B": pa.chunked_array([], pa.timestamp(unit="ns")),
                    "C": pa.chunked_array([], pa.int64()),
                }
            ),
        ),
    )


def test_difference_days():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array(
                        [dt(2019, 1, 1), dt(2020, 3, 2)], pa.timestamp(unit="ns")
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1), dt(2020, 1, 2)], pa.timestamp(unit="ns")
                    ),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="day",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array(
                        [dt(2019, 1, 1), dt(2020, 3, 2)], pa.timestamp(unit="ns")
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1), dt(2020, 1, 2)], pa.timestamp(unit="ns")
                    ),
                    "C": pa.array([365.0, -60.0]),
                }
            ),
        ),
    )


def test_difference_reuse_outcolname():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array(
                        [dt(2019, 1, 1), dt(2020, 3, 2)], pa.timestamp(unit="ns")
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1), dt(2020, 1, 2)], pa.timestamp(unit="ns")
                    ),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="day",
                outcolname="A",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "B": pa.array(
                        [dt(2020, 1, 1), dt(2020, 1, 2)], pa.timestamp(unit="ns")
                    ),
                    "A": pa.array([365.0, -60.0]),  # appended to table
                }
            ),
        ),
    )


def test_difference_fractional():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array(
                        [dt(2019, 1, 1, 0), dt(2020, 3, 2, 0, 0)],
                        pa.timestamp(unit="ns"),
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1, 12), dt(2020, 1, 2, 0, 45)],
                        pa.timestamp(unit="ns"),
                    ),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="day",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array(
                        [dt(2019, 1, 1, 0), dt(2020, 3, 2, 0, 0)],
                        pa.timestamp(unit="ns"),
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1, 12), dt(2020, 1, 2, 0, 45)],
                        pa.timestamp(unit="ns"),
                    ),
                    "C": pa.array([365.5, -59.96875]),
                }
            ),
        ),
    )


def test_difference_seconds():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array(
                        [dt(2019, 1, 1, 12, 1, 1), dt(2020, 3, 2, 8, 1, 1)],
                        pa.timestamp(unit="ns"),
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1, 2, 3, 4), dt(2020, 1, 2, 1, 2, 3)],
                        pa.timestamp(unit="ns"),
                    ),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="second",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array(
                        [dt(2019, 1, 1, 12, 1, 1), dt(2020, 3, 2, 8, 1, 1)],
                        pa.timestamp(unit="ns"),
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1, 2, 3, 4), dt(2020, 1, 2, 1, 2, 3)],
                        pa.timestamp(unit="ns"),
                    ),
                    "C": pa.array([31500123.0, -5209138.0]),
                }
            ),
        ),
    )


def test_difference_nulls():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array(
                        [None, dt(2020, 3, 2), None], pa.timestamp(unit="ns")
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1), None, None], pa.timestamp(unit="ns")
                    ),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="day",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array(
                        [None, dt(2020, 3, 2), None], pa.timestamp(unit="ns")
                    ),
                    "B": pa.array(
                        [dt(2020, 1, 1), None, None], pa.timestamp(unit="ns")
                    ),
                    "C": pa.array([None, None, None], pa.float64()),
                }
            ),
        ),
    )


def test_difference_nanoseconds():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array(
                        [1237342345234234234, 1230000034123123423, None, None],
                        pa.timestamp("ns"),
                    ),
                    "B": pa.array(
                        [1237343345234134214, 1230080234113143429, 123, None],
                        pa.timestamp("ns"),
                    ),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="nanosecond",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array(
                        [1237342345234234234, 1230000034123123423, None, None],
                        pa.timestamp("ns"),
                    ),
                    "B": pa.array(
                        [1237343345234134214, 1230080234113143429, 123, None],
                        pa.timestamp("ns"),
                    ),
                    "C": pa.array([999999899980, 80199990020006, None, None]),
                }
            ),
        ),
    )


def test_difference_nanoseconds_null():
    assert_result_equals(
        render(
            pa.table(
                {
                    "A": pa.array([123, None, None], pa.timestamp("ns")),
                    "B": pa.array([None, 123, None], pa.timestamp("ns")),
                }
            ),
            P(
                operation="difference",
                colname1="A",
                colname2="B",
                unit="nanosecond",
                outcolname="C",
            ),
        ),
        ArrowRenderResult(
            pa.table(
                {
                    "A": pa.array([123, None, None], pa.timestamp("ns")),
                    "B": pa.array([None, 123, None], pa.timestamp("ns")),
                    "C": pa.array([None, None, None], pa.int64()),
                }
            ),
        ),
    )


def test_startof_hour():
    assert_result_equals(
        render(
            make_table(
                make_column(
                    "A",
                    [
                        dt(2021, 5, 5, 13, 1, 22, 321231),
                        dt(1800, 1, 1, 1, 2, 3, 4),
                        None,
                    ],
                )
            ),
            P(operation="startof", colnames=["A"], roundunit="hour"),
        ),
        ArrowRenderResult(
            make_table(make_column("A", [dt(2021, 5, 5, 13), dt(1800, 1, 1, 1), None]))
        ),
    )


def test_startof_out_of_bounds():
    result = render(
        make_table(
            make_column(
                "A",
                [dt(1970, 1, 1), dt(1677, 9, 21, 0, 12, 43, 145500)],
            )
        ),
        P(operation="startof", colnames=["A"], roundunit="minute"),
    )
    print(repr(result.table["A"]))
    assert_result_equals(
        render(
            make_table(
                make_column(
                    "A",
                    [dt(1970, 1, 1), dt(1677, 9, 21, 0, 12, 43, 145500)],
                )
            ),
            P(operation="startof", colnames=["A"], roundunit="minute"),
        ),
        ArrowRenderResult(
            make_table(make_column("A", [dt(1970, 1, 1), None])),
            [
                RenderError(
                    i18n_message(
                        "warning.convertedOutOfBoundsToNull",
                        {"timestamp": "1677-09-21T00:12Z"},
                    )
                )
            ],
        ),
    )


tests/test_migrate_params.py
from pathlib import Path

from cjwmodule.spec.testing import param_factory

from timestampmath import migrate_params

P = param_factory(Path(__file__).parent.parent / "timestampmath.yaml")


def test_v0_to_v1():
    assert migrate_params(
        dict(
            operation="difference",
            colnames=["A", "B"],
            colname1="C",
            colname2="D",
            unit="minute",
            outcolname="froop",
        )
    ) == P(
        operation="difference",
        colnames=["A", "B"],
        colname1="C",
        colname2="D",
        unit="minute",
        roundunit="hour",
        outcolname="froop",
    )


timestampmath.py
from typing import List, NamedTuple

import numpy as np
import pyarrow as pa
import pyarrow.compute
from cjwmodule.arrow.types import ArrowRenderResult
from cjwmodule.i18n import trans
from cjwmodule.types import RenderError

_NS_PER_UNIT = {
    "nanosecond": 1,
    "microsecond": 1000,
    "millisecond": 1000000,
    "second": 1000000000,
    "minute": 60 * 1000000000,
    "hour": 3600 * 1000000000,
    "day": 86400 * 1000000000,
}


def migrate_params(params):
    if "roundunit" not in params:
        params = _migrate_params_v0_to_v1(params)
    return params


def _migrate_params_v0_to_v1(params):
    """v0 has no roundunit. v1 has it, default="hour"."""
    return {**params, "roundunit": "hour"}


def _render_minimum_or_maximum(table, colnames, outcolname, fn):
    if not colnames:
        return ArrowRenderResult(table)

    out_np_arrays = []

    num_chunks = table[colnames[0]].num_chunks
    for chunk in range(num_chunks):
        in_np_arrays = [
            table[colname].chunk(chunk).to_numpy(zero_copy_only=False)
            for colname in colnames
        ]
        out_np_array = fn.reduce(in_np_arrays)
        out_np_arrays.append(out_np_array)

    if outcolname in table.column_names:
        table = table.remove_column(table.column_names.index(outcolname))

    table = table.append_column(
        outcolname, pa.chunked_array(out_np_arrays, pa.timestamp("ns"))
    )
    return ArrowRenderResult(table)


def _render_maximum(table, colnames, outcolname):
    return _render_minimum_or_maximum(table, colnames, outcolname, np.fmax)


def _render_minimum(table, colnames, outcolname):
    return _render_minimum_or_maximum(table, colnames, outcolname, np.fmin)


def _render_difference(table, colname1, colname2, unit, outcolname):
    if not colname1 or not colname2:
        return ArrowRenderResult(table)

    out_arrays = []
    if unit == "nanosecond":
        out_type = pa.int64()
    else:
        out_type = pa.float64()
    num_chunks = table[colname1].num_chunks
    for chunk in range(num_chunks):
        in_np_array1 = table[colname1].chunk(chunk).to_numpy(zero_copy_only=False)
        in_np_array2 = table[colname2].chunk(chunk).to_numpy(zero_copy_only=False)
        out_np_timedelta_array = in_np_array2 - in_np_array1
        if unit == "nanosecond":
            # Nanosecond differences are integers
            out_array = pa.array(
                out_np_timedelta_array.astype("int"),
                mask=np.isnat(out_np_timedelta_array),
            )
        else:
            out_array = pa.array(
                out_np_timedelta_array.astype("float64") / _NS_PER_UNIT[unit],
                mask=np.isnat(out_np_timedelta_array),
            )
        out_arrays.append(out_array)

    if outcolname in table.column_names:
        table = table.remove_column(table.column_names.index(outcolname))

    table = table.append_column(outcolname, pa.chunked_array(out_arrays, out_type))

    return ArrowRenderResult(table)


class StartofColumnResult(NamedTuple):
    column: pa.ChunkedArray
    truncated: bool


def _startof(column: pa.ChunkedArray, unit: str) -> StartofColumnResult:
    factor = pa.scalar(_NS_PER_UNIT[unit], pa.int64())
    timestamp_ints = column.cast(pa.int64())

    # In two's complement, truncation rounds _up_. Subtract before truncating.
    #
    # In decimal, if we're truncating to the nearest 10:
    #
    # 0 => 0
    # -1 => -10
    # -9 => -10
    # -10 => -10
    # -11 => -20
    #
    # ... rule is: subtract 9 from all negative numbers, then truncate.

    negative = pa.compute.less(timestamp_ints, pa.scalar(0, pa.int64()))
    # "offset": -9 for negatives, 0 for others
    offset = pa.compute.multiply(
        negative.cast(pa.int64()), pa.scalar(-1 * _NS_PER_UNIT[unit] + 1, pa.int64())
    )
    # to_truncate may overflow; in that case, to_truncate > timestamp_ints
    to_truncate = pa.compute.add(timestamp_ints, offset)
    truncated = pa.compute.multiply(pa.compute.divide(to_truncate, factor), factor)

    # Mask of [True, None, True, True, None]
    safe_or_null = pa.compute.or_kleene(
        pa.compute.less_equal(to_truncate, timestamp_ints), pa.scalar(None, pa.bool_())
    )

    truncated_or_null = truncated.filter(
        safe_or_null, null_selection_behavior="emit_null"
    )

    return StartofColumnResult(
        column=truncated_or_null.cast(pa.timestamp("ns")),
        truncated=(truncated_or_null.null_count > column.null_count),
    )


def _out_of_bounds_timestamp(unit: str):
    return {
        "microsecond": "1677-09-21T00:12:43.145224Z",
        "millisecond": "1677-09-21T00:12:43.145Z",
        "second": "1677-09-21T00:12:43Z",
        "minute": "1677-09-21T00:12Z",
        "hour": "1677-09-21T00:00Z",
    }[unit]


def _render_startof(
    table: pa.Table, colnames: List[str], unit: str
) -> ArrowRenderResult:
    truncated = False
    for colname in colnames:
        i = table.column_names.index(colname)
        column_result = _startof(table.columns[i], unit)
        table = table.set_column(i, colname, column_result.column)
        if column_result.truncated:
            truncated = True

    if truncated:
        errors = [
            RenderError(
                trans(
                    "warning.convertedOutOfBoundsToNull",
                    "Converted timestamp {timestamp} to null because it is out of bounds.",
                    {"timestamp": _out_of_bounds_timestamp(unit)},
                )
            )
        ]
    else:
        errors = []

    return ArrowRenderResult(table, errors=errors)


def render_arrow_v1(table: pa.Table, params, **kwargs):
    operation = params["operation"]

    if operation in {"minimum", "maximum", "difference"} and not params["outcolname"]:
        # TODO warning? Default parameter?
        return ArrowRenderResult(table)

    if operation == "minimum":
        return _render_minimum(table, params["colnames"], params["outcolname"])
    elif operation == "maximum":
        return _render_maximum(table, params["colnames"], params["outcolname"])
    elif operation == "startof":
        return _render_startof(table, params["colnames"], params["roundunit"])
    else:
        return _render_difference(
            table,
            params["colname1"],
            params["colname2"],
            params["unit"],
            params["outcolname"],
        )
